ddtrace/contrib/botocore/services/kinesis.py

```diff
@@ -36,56 +36,69 @@ class TraceInjectionSizeExceed(Exception):
         return
 
-    data = record["Data"]
-    line_break, data_obj = get_kinesis_data_object(data)
-    if data_obj is not None:
-        data_obj["_datadog"] = {}
-        HTTPPropagator.inject(span.context, data_obj["_datadog"])
-        data_json = json.dumps(data_obj)
-
-        # if original string had a line break, add it back
-        if line_break is not None:
-            data_json += line_break
-
-        # check if data size will exceed max size with headers
-        data_size = len(data_json)
-        if data_size >= MAX_KINESIS_DATA_SIZE:
-            raise TraceInjectionSizeExceed(
-                "Data including trace injection ({}) exceeds ({})".format(data_size, MAX_KINESIS_DATA_SIZE)
-            )
+    # always inject if Data Stream is enabled, otherwise only inject if distributed tracing is enabled and this is the
+    # first record in the payload
+    if (config.botocore["distributed_tracing"] and inject_trace_context) or config._data_streams_enabled:
+        data = record["Data"]
+        line_break, data_obj = get_kinesis_data_object(data)
+        if data_obj is not None:
+            data_obj["_datadog"] = {}
+
+            if config.botocore["distributed_tracing"] and inject_trace_context:
+                HTTPPropagator.inject(span.context, data_obj["_datadog"])
+
+            core.dispatch("botocore.kinesis.start", [stream, data_obj["_datadog"]])
+
+            data_json = json.dumps(data_obj)
+
+            # if original string had a line break, add it back
+            if line_break is not None:
+                data_json += line_break
 
-        record["Data"] = data_json
+            # check if data size will exceed max size with headers
+            data_size = len(data_json)
+            if data_size >= MAX_KINESIS_DATA_SIZE:
+                raise TraceInjectionSizeExceed(
+                    "Data including trace injection ({}) exceeds ({})".format(data_size, MAX_KINESIS_DATA_SIZE)
+                )
 
+            record["Data"] = data_json
 
-def inject_trace_to_kinesis_stream(params, span):
-    # type: (List[Any], Span) -> None
+
+def inject_trace_to_kinesis_stream(params, span, inject_trace_context=True):
+    # type: (List[Any], Span, bool) -> None
     """
     :params: contains the params for the current botocore action
     :span: the span which provides the trace context to be propagated
+    :inject_trace_context: whether to inject DataDog trace context
     Max data size per record is 1MB (https://aws.amazon.com/kinesis/data-streams/faqs/)
     """
-    core.dispatch("botocore.kinesis.start", [params])
+    stream = params.get("StreamARN", params.get("StreamName", ""))
     if "Records" in params:
         records = params["Records"]
 
         if records:
-            record = records[0]
-            inject_trace_to_kinesis_stream_data(record, span)
+            for i in range(0, len(records)):
+                inject_to_trace = inject_trace_context and i == 0
```

Comment:
This means we only support batches, right?

Reply:
No, we will support both now. The code for a single putRecord call is further below on line 101.
```
    elif "Data" in params:
        inject_trace_to_kinesis_stream_data(params, span)
        inject_trace_to_kinesis_stream_data(params, span, stream, inject_trace_context=inject_trace_context)
```